specs/agents/tracing-spans-service-target.md

```diff
@@ -104,29 +104,32 @@ This specification assumes that values for `span.type` and `span.subtype` fit th
 - `span.context.service.target.type` should have the same value as `span.subtype` and fallback to `span.type`.
 - `span.context.service.target.name` depends on the span context attributes
 
-On agents, the following algorithm should be used to infer the values for `span.context.service.target.*` fields. 
+On agents, the following algorithm should be used to infer the values for `span.context.service.target.*` fields.
+
 ```javascript
 // span created on agent
 span = {};
 
 if (span.isExit) {
   service_target = span.context.service.target;
-  
-  if (!service_target.type) { // infer type from span type & subtype
-      
-    // use sub-type if provided, fallback on type othewise
+
+  if ('type' in service_target) { // If not manually specified, infer type from span type & subtype.
     service_target.type = span.subtype || span.type;
   }
-  
-  if (!service_target.name) { // infer name from span attributes
-      
-    if (span.context.db.instance) {  // database spans
-      service_target.name = span.context.db.instance;
+
+  if ('name' in service_target) { // If not manually specified, infer name from span attributes.
+
+    if (span.context.db) {  // database spans
+      if (span.context.db.instance) {
+        service_target.name = span.context.db.instance;
+      }
```

Comment:
Q2: [This spec change](https://github.com/elastic/apm/pull/670) makes it so an Elasticsearch span can have both `context.db` *and* `context.http.url`. In the case where `context.db.instance` cannot be determined (the DB spec says to use "Cluster name, if available"), would we want the service target to be:

```
service_target = { type: 'elasticsearch' }
```

or 

```
service_target = { type: 'elasticsearch', name: 'localhost:9200' }
```

?
My assumption is that we want the first one, but the current pseudo-code would result in the second.

----

Another option might be to consider basing this on `span.type`, e.g.:

```js
switch (span.type) {
  case 'db':
    if (context.db?.instance) {
      service_target.name = context.db.instance;
    }
    break;
  case 'messaging':
    if (context.message?.queue?.name) {
      service_target.name = context.message.queue.name;
    }
    break;
  default:
    if (context.http?.url) {
      // ...
    }
}
```

Was that considered in the earlier spec discussions?

Reply:
That's a very good point here, without thinking too much I think that the first option seems to be preferable here as it avoids to commit to a name based on HTTP URL that would be made obsolete when we can capture the cluster name later on.

On the other hand, a call to ES server would be an _exit span_, which means any underlying HTTP call would be hidden and not captured as a child span of the ES span. In this case capturing the HTTP URL in `context.http.url` of the call would still make sense, same would happen when instrumenting calls to `S3` that relies on HTTP. Could we get this working as expected by making the heuristic on `context.http.url` applied last and only if no other heuristic was applied ?

Having the heuristic rely on `type` (and maybe `subtype`) was definitely an option early on, and I don't recall all the details why we discarded this option. One of them was that the span `type` and `subtype` specification isn't enforced on all agents and we still have to deal with existing agents, another one is that this heuristic needs to also be applied consistently on apm-server side as well and relying on individual fields like `context.db.instance` seemed more simple.

Reply:
> Could we get this working as expected by making the heuristic on context.http.url applied last and only if no other heuristic was applied ?

My making this a *PR* that both proposes concrete spec changes *and* asks questions is perhaps making this confusing. The current PR diff proposes the following to make it work:

https://github.com/elastic/apm/pull/674/files#diff-78ae87d6c58be8cc0ed40d4dcd78faa83a58e462ad0b836912194a5cf8aed336L123-R125

I.e. If the span has `context.db`, then use the logic for a database span, whether or not it has a `context.db.instance`.
Does this seem reasonable?
I based this on the current pseudo-code in the spec for `context.destination.service.resource`:
https://github.com/elastic/apm/blob/main/specs/agents/tracing-spans-destination.md#destination-resource